### DISCUSSION TITLE
betweenness_centrality normalization

### DISCUSSION THREAD

**Discussion by ntallar** [networkx/networkx#5207] [Q&A]
Hey!

I'm currently using the betweenness_centrality for a weighted undirected graph as:
```python
# Given G an undirected graph with weights on it's edges labelled as "weight"
nx.betweenness_centrality(G, weight="weight")
```

As the normalized parameter is by default in True I was expecting the result to be in the interval [0, 1], however that's not the case as I'm getting results as 1.946798 (the example I'm using is a bit big to share it through here).

I dug a bit into the networkx code to see if maybe I was misinterpreting something, and I found something odd on the scaling ([source](https://github.com/networkx/networkx/blob/networkx-2.6.3/networkx/algorithms/centrality/betweenness.py#L355)):
```python
def _rescale(betweenness, n, normalized, directed=False, k=None, endpoints=False):
    if normalized:
        if endpoints:
            if n < 2:
                scale = None  # no normalization
            else:
                # Scale factor should include endpoint nodes
                scale = 1 / (n * (n - 1))
        elif n <= 2:
            scale = None  # no normalization b=0 for all nodes
        else:
            scale = 1 / ((n - 1) * (n - 2))
    else:  # rescale by 2 for undirected graphs
        if not directed:
            scale = 0.5
        else:
            scale = None
    if scale is not None:
        if k is not None:
            scale = scale * n / k
        for v in betweenness:
            betweenness[v] *= scale
    return betweenness
```

According to the documentation the scaling should be `2/((n-1)(n-2))` when `normalized=True` and the graph is undirected, but that doesn't seem to be the case, whether the graph is directed is only taken into account when normalization is turned off.

I'm not familiar enough with the interface and implementation so I'm unsure of what conclusion to reach, maybe the normalization doesn't work as I initially expected. Any thoughts on this?

Cheers,
Nico

**Comment by dschult** [comment]
A little further down in the documentation for `betweenness_centrality` it says:
```
    The total number of paths between source and target is counted
    differently for directed and undirected graphs. Directed paths
    are easy to count. Undirected paths are tricky: should a path
    from "u" to "v" count as 1 undirected path or as 2 directed paths?
    For betweenness_centrality we report the number of undirected
    paths when G is undirected.
```

When you look at the code, you see the scale factor `1/((n-1)(n-2))`, but the scale factor doesn't tell you the normalization factor because it doesn't say what the numerator is counting. That counting system is harder to parse (for performance reasons), but it is counting in a way that gives twice the number of undirected paths. So the same scale factor is used for directed and undirected graphs.  That is why we have to scale by 0.5 when normalization is False.
You can also check this by looking at simple examples like a 4 node path or a small complete graph.

That does not explain why you are getting numbers greater than 1. You should get values that are in [0, 1] when the normalized option is True.
